### ISSUE TITLE
Bad AST formatting for INTERVAL and OPERATOR(*)

### ISSUE THREAD

**Issue by def-** [MaterializeInc/materialize#20110]
### What version of Materialize are you using?

main

### What is the issue?

I noticed in https://github.com/MaterializeInc/materialize/pull/19780#issuecomment-1601104025 that `SELECT INTERVAL '1:2:31.23456789' MINUTE TO SECOND(2);` is wrongly printed out again as `SELECT INTERVAL '1:2:31.23456789' SECOND (2);` by the AST printer. @vmarcos found the relevant location: https://github.com/MaterializeInc/materialize/blob/ad6d3197ef88f4d067c40532bc2eb810008ace5b/src/sql-parser/src/ast/defs/value.rs#L78-L126

Second case: `select 2 OPERATOR(*) 2 + 2` becomes `SELECT 2 * 2 + 2`.

I'm not sure if there is a case to trigger this problem straight from SQL (I'd be interested to see the query if it is), but they seem to be bugs in AstDisplay. I haven't managed to trigger it with views/mvs since the parsed SQL was always put into a different form and not emitted again using AstDisplay. But maybe I'm missing something.

**Comment by sploiselle** [comment]
```
materialize=> CREATE MATERIALIZED VIEW x AS SELECT INTERVAL '1:2:31.23456789' MINUTE TO SECOND(2);
CREATE MATERIALIZED VIEW
materialize=> show create materialized view x;
         name         |                                                        create_sql                                                        
----------------------+--------------------------------------------------------------------------------------------------------------------------
 materialize.public.x | CREATE MATERIALIZED VIEW "materialize"."public"."x" IN CLUSTER "default" AS SELECT INTERVAL '1:2:31.23456789' SECOND (2)
(1 row)

materialize=> select * from x;
  interval   
-------------
 00:02:31.23
(1 row)
```

Restart:

```
materialize=> show create materialized view x;
         name         |                                                        create_sql                                                        
----------------------+--------------------------------------------------------------------------------------------------------------------------
 materialize.public.x | CREATE MATERIALIZED VIEW "materialize"."public"."x" IN CLUSTER "default" AS SELECT INTERVAL '1:2:31.23456789' SECOND (2)
(1 row)

materialize=> SELECT INTERVAL '1:2:31.23456789' SECOND (2);
  interval   
-------------
 01:02:31.23
(1 row)

materialize=> create materialized view y AS SELECT INTERVAL '1:2:31.23456789' SECOND (2);
CREATE MATERIALIZED VIEW
materialize=> select * from y;
  interval   
-------------
 01:02:31.23
 ```
 
 This is bad because `CREATE`/`SHOW CREATE` should roundtrip.

**Comment by def-** [comment]
Ok, great. I should have done a restart in between, that explains it! We should also add these to platform-checks when they are fixed, but I can do that too. It seems like these strange syntax constructs are generally dangerous and each of them should be validated through platform-checks.

**Comment by sploiselle** [comment]
Ah, sorry––I was showing here that the restart doesn't do anything (which confuses me because it suggests that there is some divergence between the output of `SHOW CREATE` and the string we use to create the view). But these things are bad and should be fixed.

**Comment by def-** [comment]
Alright, same effect for `OPERATOR(*)`:
```
materialize=> create view z as select 2 OPERATOR(*) 2 + 2;
CREATE VIEW
materialize=> show create view z;
         name         |                         create_sql
----------------------+------------------------------------------------------------
 materialize.public.z | CREATE VIEW "materialize"."public"."z" AS SELECT 2 * 2 + 2
(1 row)
```

**Comment by sploiselle** [comment]
Also to note that `OPERATOR(*)` has a different precedence than `*`, which means we would need to introduce some kind of...

```rust
struct Operator {
  val: String,
  // Value was surrounded by the `OPERATOR` keyword,
  // which alters its precedence
  literal: bool,
}
```

And then do a bunch of jiggery-pokey to make that work.

I think this is much less egregious/serious bug than the malformed interval literals.

**Comment by vmarcos** [comment]
Another instance of this problem can be observed with `AS OF` clauses, e.g., by running `test/sqllogictest/as_of.slt` or `test/sqllogictest/temporal.slt` with `--auto-index-selects`.

**Comment by def-** [comment]
Good timing, I'm just adding explicit tests for all the missing AstDisplay coverage: https://github.com/MaterializeInc/materialize/issues/20127

**Comment by def-** [comment]
@vmarcos Where do you see the problem? What I see in temporal.slt looks unrelated but potentially concerning. Are temporals supposed to work the same here in mv as in one-shot selects?:
```
    select * from valid AS OF 2;
    InconsistentViewOutcome:test/sqllogictest/temporal.slt:27
            expected from query: Success
            actually from indexed view: OutputFailure { expected_output: Values(["0", "10", "1", "10", "1", "2", "2", "13"]), actual_raw_output: [Row { columns: [Column { name: "a", type: Int4 }, Column { name: "b", type: Numeric }] }], actual_output: Values(["5", "18446744073709551614"]), location: Location { file: "test/sqllogictest/temporal.slt", line: 27 } }

    ----
    select * from valid AS OF 3;
    InconsistentViewOutcome:test/sqllogictest/temporal.slt:35
            expected from query: Success
            actually from indexed view: OutputFailure { expected_output: Values(["0", "10", "1", "10", "2", "13"]), actual_raw_output: [Row { columns: [Column { name: "a", type: Int4 }, Column { name: "b", type: Numeric }] }], actual_output: Values(["5", "18446744073709551614"]), location: Location { file: "test/sqllogictest/temporal.slt", line: 35 } }

    ----
    select * from valid AS OF 11;
    InconsistentViewOutcome:test/sqllogictest/temporal.slt:42
            expected from query: Success
            actually from indexed view: OutputFailure { expected_output: Values(["2", "13", "5", "18446744073709551614"]), actual_raw_output: [Row { columns: [Column { name: "a", type: Int4 }, Column { name: "b", type: Numeric }] }], actual_output: Values(["5", "18446744073709551614"]), location: Location { file: "test/sqllogictest/temporal.slt", line: 42 } }

    ----
    FAIL: inconsistent-view-outcome=3 success=11 total=14
FAIL: inconsistent-view-outcome=3 success=11 total=14
```
At least the query itself can roundtrip in the parser:
```
parse-statement
SELECT * FROM valid AS OF 2
----
SELECT * FROM valid AS OF 2
=>
Select(SelectStatement { query: Query { ctes: Simple([]), body: Select(Select { distinct: None, projection: [Wildcard], from: [TableWithJoins { relation: Table { name: Name(UnresolvedItemName([Ident("valid")])), alias: None }, joins: [] }], selection: None, group_by: [], having: None, options: [] }), order_by: [], limit: None, offset: None }, as_of: Some(At(Value(Number("2")))) })
```

**Comment by vmarcos** [comment]
@def- If we run `sqllogictest` with both `-vv` and `--auto-index-selects`, then we see output like the following for `temporal.slt`:
```
    select * from valid AS OF 2;
    CREATE VIEW "v962a18b4a1b54fd6b63284ea2b1014a8" AS SELECT * FROM "valid"
    CREATE DEFAULT INDEX ON "v962a18b4a1b54fd6b63284ea2b1014a8"
    SELECT * FROM "v962a18b4a1b54fd6b63284ea2b1014a8"
    DROP VIEW "v962a18b4a1b54fd6b63284ea2b1014a8"
    InconsistentViewOutcome:test/sqllogictest/temporal.slt:27
            expected from query: Success
            actually from indexed view: OutputFailure { expected_output: Values(["0", "10", "1", "10", "1", "2", "2", "13"]), actual_raw_output: [Row { columns: [Column { name: "a", type: Int4 }, Column { name: "b", type: Numeric }] }], actual_output: Values(["5", "18446744073709551614"]), location: Location { file: "test/sqllogictest/temporal.slt", line: 27 } }
```
So the `CREATE VIEW` statement comes out without the `AS OF` specification for some reason. It should instead have been a parse error:
```sql
materialize=> create view v as select * from valid AS OF 2;
ERROR:  Expected end of statement, found AS
LINE 1: create view v as select * from valid AS OF 2;
                                             ^
Time: 0.398 ms
```
Perhaps this is evading the roundtrip test because it is a direct AST transformation? I take the parsed `SELECT` and assemble a `CREATE VIEW` statement with it. I'll dig into it.

**Comment by vmarcos** [comment]
> Perhaps this is evading the roundtrip test because it is a direct AST transformation? I take the parsed SELECT and assemble a CREATE VIEW statement with it. I'll dig into it.

Indeed, this was my bad; `AS OF` clauses were not being considered correctly by `sqllogictest`'s `--auto-index-selects`. A proposal for a fix is now in #20504.

**Comment by def-** [comment]
This should be all fixed based on what I tested.
